Improved Documentation for Contributing To Notification Center (#14475)
meshery/meshery: docs/pages/project/contributing/contributing-ui-notification-center.md

- **metadata.js**: Defines `PropertyFormatters`, `LinkFormatters`, `PropertyLinkFormatters`, and `EventTypeFormatters`. Contains the `FormattedMetadata` component which decides _how_ to format the metadata based on event type or specific properties.
- **notification.js**: Defines how an individual notification is rendered.

#### `formatters/` _(NotificationCenter/formatters)_

This directory houses reusable formatter components dedicated to specific types of metadata or event types.

- **common.js**: Contains shared components like `TitleLink`, `DataToFileLink`, `EmptyState`.
- **error.js**: Defines `ErrorMetadataFormatter` for displaying structured error details.
- **model_registration.js**: Contains formatters for model import/registration events (`ModelImportMessages`, `ModelImportedSection`).
- **pattern_dryrun.js**: Defines `DryRunResponseFormatter` which utilizes components from `DesignLifeCycle`.
- **relationship_evaluation.js**: Defines `RelationshipEvaluationEventFormatter` responsible for rendering notifications related to the evaluation of relationships between components in a design.

> ### Types of Event Specific Notification Formatters

Reviewer: ```suggestion
## Types of Event Specific Notification Formatters
```

> ## Common Formatter

Reviewer: ```suggestion
### Common Formatter
```

The following reusable components standardize how notification links, empty states, and downloadable traces are displayed:

1. **TitleLink**: Renders a styled title with an external link icon.  
   Props:

   - `href` (required): URL of the link.
   - `children`: The link text.

2. **EmptyState**: Displays a description when no specific data is available for an event.  
   Props:

   - `event.description` (optional): Text description of the event.

3. **DataToFileLink**: Converts event data into a downloadable `.txt` file.  
   Props:
   - `data` (required): Can be a string or JSON object.

> ## Error Formatter

Reviewer: ```suggestion
### Error Formatter
```

  - `LongDescription`: Provides details about the error.
  - `ProbableCause`: Lists possible reasons for the error.
  - `SuggestedRemediation`: Suggests solutions to fix the error.
- `event` (object, optional): Contains the notification event data.

**Path:** `ui/components/NotificationCenter/formatters/error.js`

**Example:**

```javascript
<ErrorMetadataFormatter
  metadata={
    LongDescription: "An unexpected error occurred while deploying the mesh.",
    ProbableCause: "Misconfigured Kubernetes cluster.",
    SuggestedRemediation: "Check your kubeconfig file and retry deployment.",
  }
  event={ description: "Mesh deployment failed" }
/>
```

<a href="{{ site.baseurl }}/assets/img/notification-center/error-formatter.png"><img alt="Error Formatter" style="width:500px;height:auto;" src="{{ site.baseurl }}/assets/img/notification-center/error-formatter.png" /></a>

**When to Use:**

The `ErrorMetadataFormatter` is used when dealing with structured error events that follow a pattern (description, cause, remediation). A new formatter should be created only if the error metadata deviates significantly from the `ErrorMetadataFormatter` metadata structure.

> ## Model Registration Formatter

Reviewer: ```suggestion
### Model Registration Formatter
```

The `Model Registration Formatter` formats and displays model registration details, including components and relationships, in Meshery UI's Notification Center. It ensures structured representation of imported models and error handling during the import process.

**Path:** `ui/components/NotificationCenter/formatters/model_registration.js`

**Components:**

1. **UnsuccessfulEntityWithError**: This component is used to handle error cases during model import. It identifies the type and count of entities that failed to import.

   **Props:**

   - `modelName` (string): The name of the model being imported.
   - `error` (object): Contains details about the error encountered during import.

2. **ModelImportedSection**: Displays the details of the imported model along with components, relationships, and any errors that occur.

   **Props:**

   - `modelDetails` (object) – Contains model import data.

<a href="{{ site.baseurl }}/assets/img/notification-center/model-register-formatter.png"><img alt="Model Register Formatter" style="width:500px;height:auto;" src="{{ site.baseurl }}/assets/img/notification-center/model-register-formatter.png" /></a>

> ## Relationship Evaluation Formatter

Reviewer: ```suggestion
### Relationship Evaluation Formatter
```

   This gives the user context about which design and version were evaluated.

2. **Detailed Changes**:  
   The notification breaks down the changes into categories:

   - **Components**:
     - **Added**: New components introduced in the design.
     - **Updated**: Existing components that were modified.
     - **Removed**: Components that were deleted from the design.
   - **Relationships**:
     - **Added**: New relationships established between components.
     - **Updated**: Existing relationships that were modified.
     - **Removed**: Relationships that were deleted.

3. **Component Details**:  
   For each component, the notification displays:

   - **Kind**: The type of the component (e.g., `Deployment`, `Service`).
   - **Name**: The name of the component.
   - **Model and Version**: The model and version associated with the component.

4. **Relationship Details**:  
   For each relationship, the notification displays:
   - **Type**: The type of relationship.
   - **Source and Target**: The components involved in the relationship (e.g., `Pod` to `Pod`).
   - **Model and Version**: The model and version associated with the relationship.

> ## Dry Run Formatter

Reviewer: ```suggestion
### Dry Run Formatter
```

1. **DeploymentSummaryFormatter**:  
   This component is used to display:
   - **Event Description**: A brief description of the deployment event.
   - **Errors**: Any errors encountered during the deployment process.
   - **Component Details**: A list of components with their deployment status and metadata.

**Props:**

- **event** (object): Contains:
  - **description**: A brief description of the deployment event.
  - **action**: The type of action performed (e.g., `deploy`, `undeploy`).
  - **severity**: The severity level of the event.
  - **metadata**:
    - **summary**: A detailed summary of the deployment process, including component details.
    - **error**: Any errors encountered during the deployment process.
    - **design_name**: The name of the design being deployed.
    - **design_id**: The ID of the design being deployed.

#### When to Use

The **Deployment Summary Formatter** should be used in the following scenarios:

1. **Deployment or Undeployment Events**:  
   When the `event.action` is either `deploy` or `undeploy` and the `event.metadata` includes `design_name`.
2. When a notification involves the deployment or undeployment of a design.

> ## PropertyFormatters and PropertyLinkFormatters

Reviewer: ```suggestion
### PropertyFormatters and PropertyLinkFormatters
```

 Your browser does not support the video tag
</video>

---

### What is the Notification Center?

The Notification Center is a dedicated panel in Meshery’s UI that helps you monitor, understand, and respond to events across your system. It acts as a central place where you can see important updates related to your infrastructure, workloads, and Meshery’s internal operations.

> 

Reviewer: ```suggestion
> Want to understand how users interact with the Notification Center? [Learn more here]({{ site.baseurl }}/guides/infrastructure-management/notification-management/user-intent).
```

> > Want to understand how users interact with the Notification Center? [Learn more here]({{ site.baseurl }}/guides/infrastructure-management/notification-management/user-intent).

Reviewer: ```suggestion

> Want to understand how users interact with the Notification Center? [Learn more here](https://docs.meshery.io/guides/infrastructure-management/notification-management).

```

- Network Request are cached and are invalidated when new events come or events are deleted/updated.
- Due to need for infinite scroll and optimistic update the events are stored globally in Redux.

> ## Bulk Operations

Reviewer: ```suggestion
### Bulk Operations
```

Bulk operations in the Notification Center allow users to perform actions like deleting multiple notifications or changing the status of multiple notifications in a batch. This documentation outlines the key features and functionality of bulk operations, including the restriction of performing only one bulk operation at a time, the disabling of buttons during ongoing operations, and the display of a loading icon to indicate ongoing activity.

> **How It Works**

Reviewer: ```suggestion
```

### Initiating a Bulk Operation

- Users select the notifications they want to include in the bulk operation. This is typically done by checking checkboxes next to each notification.
- After selecting notifications, users trigger the desired bulk operation (e.g., delete or change status) by clicking the corresponding action button.
- Once initiated, the bulk operation begins processing the selected notifications.

## Metadata Formatter

> When an event is received from the server, it adheres to a fixed schema containing information that is valuable for presentation to the user. This information typically includes details such as the description, date, user_id, system_id, action, and acted-upon resources. Additionally, sometimes there may be a detailed traceback, a summary, or a comprehensive error log, all of which are dynamically generated data encapsulated within the metadata of the event. Presenting this structured data in a user-friendly manner is a crucial task because it contains valuable insights into ongoing operations.

Reviewer: ```suggestion
When the server sends an event, it follows a consistent schema that contains metadata intended for user presentation. This metadata typically includes fields such as `description`, `date`, `user_id`, `system_id`, `action`, and the resources involved.  

In some cases, the metadata may also contain more detailed information—such as a traceback, a summary, or a complete error log—which is dynamically generated at runtime and encapsulated within the event.  

Presenting this structured information in a clear and accessible way is essential, as it provides valuable insights into system behavior and ongoing operations.
```

1. **Deployment or Undeployment Events**:  
   When the `event.action` is either `deploy` or `undeploy` and the `event.metadata` includes `design_name`.
2. When a notification involves the deployment or undeployment of a design.

---

### PropertyFormatters and PropertyLinkFormatters

**Purpose:**  
When an event does not match an event in `EventTypeFormatters`, **PropertyFormatters** and **PropertyLinkFormatters** are used to format and render specific metadata fields in a structured and visually appealing way.

---

#### Examples of Property Formatters

1. **trace**: Converts large trace data into a downloadable file link.
2. **ShortDescription**: Displays a short description of the event.
3. **Error**: Uses the `ErrorMetadataFormatter` to display structured error details.

---

#### When to Use

Use **PropertyFormatters** and **PropertyLinkFormatters** in the following scenarios:

1. When an event does not have a specific `EventTypeFormatter` defined.
2. When you need to render individual metadata fields in a structured and visually appealing way.
3. When metadata includes fields like trace data, short descriptions, or error details that require specialized formatting.

---

> ### How to Add a New EventType Formatter(?)

Reviewer: ```suggestion
```
